proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1205 @@
+    handle<$resource> => {
+      val = values.assert_next(i32)
+      yield(TABLES.get($resource, val, src) as handle<$resource>)
+
+      # If we're lifting for an exported function then we're lifting the results
+      # of the export after the call. Similar to lists the handle index
+      # reference is owned by the callee (which just returned) but is a sort of
+      # "temporary" allocation that the callee doesn't need. As a result, like
+      # lists, we automatically remove the index from the source's index space.
+      if direction == "export":
+        TABLES.remove($resource, val, src)
+    }
+
+    # Buffers are only lifted from imports, which means that this only applies
+    # to the parameters of imported functions. This means that the
+    # representation we're dealing with is three i32 values.
+    #
+    # The first i32 is either 0 meaning that this is a new local buffer within
+    # `src`, or it's a nonzero value meaning that it's a handle to a buffer
+    # defined in some other module.
+    #
+    # If the first handle value is 0 then the offset parameter is in bytes from
+    # the start of the memory of `src`. The `len` parameter is the length, in
+    # units of $ty, of the buffer. The pointer must be well aligned.
+    #
+    # If the first handle value is not zero then TABLES.get_buffer will validate
+    # that the handle is indeed a valid buffer for `src` to reference. The
+    # `offset` and `len` parameters are in units of `$ty` relative to within the
+    # original buffer. These values are validated to ensure they fit within the
+    # original buffer
+    push-buffer<$ty> | pull-buffer<$ty> => {
+      assert(direction == "import")
+      handle = values.assert_next(i32)
+      offset = values.assert_next(i32)
+      len = values.assert_next(i32)
+      if handle == 0:
+        assert_or_trap(offset + len * size(direction, $ty) <= src.memory.len)
+        yield(Buffer(src, offset, len, ty))
+      else:
+        buffer = TABLES.get_buffer(handle, src, ty).clone()
+        assert_or_trap(offset + len <= buffer.len)
+        buffer.offset += offset * size(direction, $ty)
+        buffer.len = len
+        yield(buffer)
+    }
+```
+
+There's a few more missing pieces to fill in, namely the in-memory
+representation of types due to the representation of `list` as contiguous
+elements in-memory. For this we can first define the `size` and `align` of all
+types:
+
+```
+def size(direction, type):
+  match type
+    f32 => 4
+    f64 => 8
+    s8 => 1
+    u8 => 1
+    s16 => 2
+    u16 => 2
+    s32 => 4
+    u32 => 4
+    s64 => 8
+    u64 => 8
+    char => 4
+    string |
+    list $ty => size(record([i32, i32]))
+    flags $fields => num_packed_i32s($fields) * 4
+    record $fields =>
+      s = 0
+      for field in $fields
+        s = align_to(s, align(field))
+        s += size(direction, field)
+      align_to(s, align(record $fields))
```

Comment:
This means that empty (zst) records get size/alignment of 1. Seems like it makes sense to explicitly call out ZST behaviors (not sure if a filed with zero fileds is a thing though).

Reply:
Oh the intention is that zst types have zero size and alignment of 1. I should probably actually define the `align_to` function here, but it would make it clear that zero is aligned to all alignments, so the size here would remain 0.